### PR TITLE
Cleanup inactive transactions in non-validator

### PR DESCRIPTION
**Pull request by Kourin1996**
Fix EDGE-843

# Description

This PR adds a function to remove the transactions that have not been adopted for a certain number of transactions.
And also this PR disable `--seal` flag but keep CLI flag because this PR is intended to be merged in the next PATCH release

Relevant: https://github.com/0xPolygon/polygon-edge/issues/731

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

1. Send the multiple transactions with consecutive nonces to **non-validator node** so that validators process the first part of transactions and fail to include the second part of the transactions (e.g. Send the bigger amount than the value account holds in a multiple transactions)
2. Make sure there are pending transactions in TxPool of validator and non-validator
3. Make sure the first part of the transactions are processed
4. Wait for a while
5. Make sure there are no transactions in TxPool of validator and non-validator

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: txpool/txpool.go [hunk 1]
```diff
@@ -321,6 +323,16 @@ func (p *TxPool) SetSigner(s signer) {
 	p.signer = s
 }
 
+// SetSealing sets the sealing flag
+func (p *TxPool) SetSealing(sealing bool) {
+	p.sealing.Store(sealing)
+}
+
+// Sealing returns the current set sealing flag
+func (p *TxPool) Sealing() bool {
```
Comment (dbrajovic) [inline txpool/txpool.go]: You can un-export this method

File: txpool/account.go [hunk 2]
```diff
@@ -157,6 +157,7 @@ type account struct {
 	enqueued, promoted *accountQueue
 	nextNonce          uint64
 	demotions          uint
+	unadopted          uint64
```
Comment (dbrajovic) [inline txpool/account.go]: Naming nitpick: unadopted -> skips

File: txpool/txpool.go [hunk 3]
```diff
@@ -881,6 +898,49 @@ func (p *TxPool) resetAccounts(stateNonces map[types.Address]uint64) {
...
+	p.accounts.Range(
+		func(key, value interface{}) bool {
+			address, _ := key.(types.Address)
+			account, _ := value.(*account)
+
+			if _, ok := latestActiveAccounts[address]; ok {
+				account.unadopted = 0
+			} else {
+				account.unadopted++
+			}
```
Comment (dbrajovic) [inline txpool/txpool.go]: Nitpick: make this a method of `account`

File: txpool/txpool.go [hunk 4]
```diff
@@ -881,6 +898,49 @@ func (p *TxPool) resetAccounts(stateNonces map[types.Address]uint64) {
...
+
+			account.enqueued.lock(true)
+			defer account.enqueued.unlock()
+
+			firstTx = account.enqueued.pop()
+			if firstTx != nil {
+				p.index.remove(firstTx)
+				p.gauge.decrease(slotsRequired(firstTx))
+				account.unadopted = 0
+			}
```
Comment (dbrajovic) [inline txpool/txpool.go:937]: This is the critical part the txpool needs to do correctly: since we're trying to mimic what validator nodes do when performing the `Demote->Drop` this piece of code should be replaced with a call to `Drop()` 

(Drop does all of this logic already and updates the metrics as well)

The core idea is symmetrical - `demotions` field is for validators `skips` (unadopted) for non-validators, but the behavior needs to be identical.

Comment (dbrajovic) [inline txpool/txpool.go]: This is the critical part the txpool needs to do correctly: since we're trying to mimic what validator nodes do when performing the `Demote->Drop`. This piece of code should be replaced with a call to `Drop()` 

(Drop does all of this logic already and updates the metrics as well)

The core idea is symmetrical - `demotions` field is for validators, `skips` (unadopted) for non-validators, but the behavior needs to be identical.

**Review by dbrajovic — CHANGES_REQUESTED** [review]
Already looks good, just need to resolve the `updateUnadoptedCounts` detail 👍

File: consensus/dev/dev.go [hunk 5]
```diff
@@ -61,6 +61,8 @@ func Factory(
 
 // Initialize initializes the consensus
 func (d *Dev) Initialize() error {
+	d.txpool.SetSealing(true)
```
Comment (zivkovicmilos) [inline consensus/dev/dev.go:64]: Why do you need this anymore?

File: consensus/ibft/ibft.go [hunk 6]
```diff
@@ -41,6 +41,7 @@ type txPoolInterface interface {
 	Drop(tx *types.Transaction)
 	Demote(tx *types.Transaction)
 	ResetWithHeaders(headers ...*types.Header)
+	SetSealing(bool)
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go:44]: You don't need this anymore

File: txpool/txpool.go [hunk 7]
```diff
@@ -321,6 +323,16 @@ func (p *TxPool) SetSigner(s signer) {
 	p.signer = s
 }
 
+// SetSealing sets the sealing flag
+func (p *TxPool) SetSealing(sealing bool) {
+	p.sealing.Store(sealing)
+}
```
Comment (zivkovicmilos) [inline txpool/txpool.go:329]: Why is this required?

**Review by zivkovicmilos — APPROVED** [review]

File: txpool/txpool.go (hunk 1 above)
Comment (Kourin1996) [reply txpool/txpool.go]: Fixed 52735acfc16c2ce6849fcc18334ad8c4c7180a0e

File: txpool/account.go (hunk 2 above)
Comment (Kourin1996) [reply txpool/account.go]: Renamed e0372e5b909ab6f543b904505f9746abfe21a078

File: txpool/txpool.go (hunk 3 above)
Comment (Kourin1996) [reply txpool/txpool.go]: Added 1f9344761d8c0676269fbf20d774528236c7f6c0

File: txpool/txpool.go (hunk 4 above)
Comment (Kourin1996) [reply txpool/txpool.go]: Fixed　9c01589800042e14d3b408732fc4249457a0b292

File: txpool/account.go [hunk 8]
```diff
@@ -293,3 +295,33 @@ func (a *account) promote() []*types.Transaction {
...
+		return firstPromoted
+	}
+
+	a.enqueued.lock(true)
+	defer a.enqueued.unlock()
+
+	if firstEnqueued := a.enqueued.peek(); firstEnqueued != nil {
+		return firstEnqueued
+	}
+
```
Comment (dbrajovic) [inline txpool/account.go:325]: This can potentially brick an account later on when `Drop` sets the nonce to that of the enqueued transaction. 

Here is my proposal on `updateAccountSkipsCounts` and `getLowestTx`:

```
	p.accounts.Range(
		func(key, value interface{}) bool {
			address, _ := key.(types.Address)
			account, _ := value.(*account)

			if _, ok := latestActiveAccounts[address]; ok {
				account.resetSkips()

				return true
			}

			firstPromoted := account.getLowestTx()
			if firstPromoted == nil {
				// no need to increment anything,
				// account has no txs
				return true
			}

			account.incrementSkips()

			if account.skips < maxAccountSkips {
				return true
			}

			// account has been skipped too many times
			p.Drop(firstPromoted)
			account.resetSkips()

			return true
		},
	)
```

```
func (a *account) getLowestTx() *types.Transaction {
	a.promoted.lock(false) // only read, no write
	defer a.promoted.unlock()

	return a.promoted.peek()
}
```

With these two, only accounts which are **skipped and contain promoted transactions** are processed. Skipped account with no transactions do not fit the criteria, so no need to increment `skips`.

Comment (Kourin1996) [reply txpool/account.go:325]: Fixed 5386a85520ccf47e44e84fbdb6f4ff9e3f4cd407

**Review by dbrajovic — APPROVED** [review]

**Review by 0xAleksaOpacic — APPROVED** [review]
